Sources/MongoSwift/ClientSession.swift

```diff
@@ -373,4 +373,146 @@ public final class ClientSession {
+     * and/or end the session. If the closure commits or aborts the current transaction and/or ends the session,
+     * however, `withTransaction` will return without taking further action.
+     *
+     * - Parameters:
+     *   - options: The options to use when starting the transaction.
+     *
+     * - Returns:
+     *   - An `EventLoopFuture<T>`, the return value of the user-provided closure.
+     *
+     *   If the future fails, the error is likely one of the following:
+     *   - `CommandError` if an error prevents a command from executing.
+     *   - `LogicError` if the session already has an in-progress transaction.
+     *   - `LogicError` if `withTransaction` is called on an ended session.
+     *
+     * - SeeAlso:
+     *   - https://docs.mongodb.com/manual/core/transactions/
+     */
+    public func withTransaction<T>(
+        options: TransactionOptions? = nil,
+        _ transactionBody: @escaping (ClientSession) throws -> EventLoopFuture<T>
+    ) -> EventLoopFuture<T> {
+        switch self.state {
+        case .notStarted, .started:
+            // Enforce a 120-second timeout to limit transaction retry behavior.
+            let retryTimeoutTime = Date(timeIntervalSinceNow: 120)
+            return self.attemptTransaction(
+                retryTimeoutTime: retryTimeoutTime,
+                options: options,
+                transactionBody
+            )
+        case .ended:
+            return self.client.operationExecutor.makeFailedFuture(ClientSession.SessionInactiveError)
+        }
+    }
+
+    // Private helper function that attempts to start a multi-document transaction, execute the provided closure, and
+    // commit the transaction. If an error occurs, the function attempts to retry the transaction, whenever possible.
+    // - Note: This function should not be used outside the context of `withTransaction`.
+    private func attemptTransaction<T>(
+        retryTimeoutTime: Date,
+        options: TransactionOptions?,
+        _ transactionBody: @escaping (ClientSession) throws -> EventLoopFuture<T>
+    ) -> EventLoopFuture<T> {
+        self.startTransaction(options: options).flatMap { _ in
+            do {
+                let transactionBodyFuture = try transactionBody(self)
+                return transactionBodyFuture.flatMap { value in
+                    if [.none, .committed, .aborted].contains(self.transactionState ?? .none) {
+                        return self.client.operationExecutor.makeSucceededFuture(value)
+                    }
+                    return self.attemptTransactionCommit(
+                        value: value,
+                        retryTimeoutTime: retryTimeoutTime,
+                        options: options,
+                        transactionBody
+                    )
+                }.flatMapError { error in
+                    if [.starting, .inProgress].contains(self.transactionState ?? .none) {
```

Comment:
If it compiles (I think it will....) I think this line would read a little clearer like
```swift
if let state = self.transactionState, [.starting, .inProgress].contains(state) {
```

Reply:
Done.